From cc785b14815ee8e85be34489c31ef2305a620f52 Mon Sep 17 00:00:00 2001
From: Junping Du <junping_du@apache.org>
Date: Wed, 11 Feb 2015 03:44:31 -0800
Subject: [PATCH 1370/2848] YARN-3160. Fix non-atomic operation on
 nodeUpdateQueue in RMNodeImpl. (Contributed by
 Chengbing Liu) (cherry picked from commit
 c541a374d88ffed6ee71b0e5d556939ccd2c5159) (cherry
 picked from commit
 f2c3058b8d57710fd0d98eb270e2f26f27aa1e3f)

Change-Id: Ia73bf1a4ad407ffb0c9d13ebff95837867209c8e
---
 .../server/resourcemanager/rmnode/RMNodeImpl.java  |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmnode/RMNodeImpl.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmnode/RMNodeImpl.java
index c534844..777f173 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmnode/RMNodeImpl.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/rmnode/RMNodeImpl.java
@@ -837,8 +837,9 @@ public NodeState transition(RMNodeImpl rmNode, RMNodeEvent event) {
   public List<UpdatedContainerInfo> pullContainerUpdates() {
     List<UpdatedContainerInfo> latestContainerInfoList = 
         new ArrayList<UpdatedContainerInfo>();
-    while(nodeUpdateQueue.peek() != null){
-      latestContainerInfoList.add(nodeUpdateQueue.poll());
+    UpdatedContainerInfo containerInfo;
+    while ((containerInfo = nodeUpdateQueue.poll()) != null) {
+      latestContainerInfoList.add(containerInfo);
     }
     this.nextHeartBeat = true;
     return latestContainerInfoList;
-- 
1.7.9.5

